### PR TITLE
Introduce ufcg for prospector

### PR DESCRIPTION
**Pull request by ruflin**
In a first step, ucfg was introduced to unpack prospector configs. In next steps the config files also for the prospector types and harvester types will be split up.

* Obsolete tests were removed

### CODE REVIEW

File: filebeat/config/config.go [hunk 1]
```diff
@@ -17,7 +15,7 @@ import (
 // Defaults for config variables which are not set
 const (
 	DefaultRegistryFile                      = "registry"
-	DefaultIgnoreOlderDuration time.Duration = 0
+	DefaultIgnoreOlder         time.Duration = 0
 	DefaultCloseOlderDuration  time.Duration = 1 * time.Hour
```
Comment (urso) [inline filebeat/config/config.go]: DefaultCloseOlder?

File: filebeat/crawler/crawler.go [hunk 2]
```diff
@@ -39,8 +39,6 @@ func (c *Crawler) Start(prospectorConfigs []config.ProspectorConfig, eventChan c
 	// Prospect the globs/paths given on the command line and launch harvesters
 	for _, prospectorConfig := range prospectorConfigs {
```
Comment (urso) [inline filebeat/crawler/crawler.go]: one can unpack here into:

```
prospectorInfo := struct {
    Type string `config:"input_type"  validate:"required"`
    Enabled bool `config:"enabled"`
}{"", true}

err := prospectorConfig.Unpack(&prospectorInfo)
...

if !prospectorInfo.Enabled {
  continue
}

plugin := lookup(prospectorInfo.Type)
```

File: filebeat/crawler/prospector.go [hunk 3]
```diff
@@ -28,9 +31,24 @@ type Prospectorer interface {
 	Run()
 }
 
-func NewProspector(prospectorConfig cfg.ProspectorConfig, registrar *Registrar, spoolerChan chan *input.FileEvent) (*Prospector, error) {
+type prospectorConfig struct {
```
Comment (urso) [inline filebeat/crawler/prospector.go]: I like to put these into separate config.go file.

Also add `func (p *prospectorConfig) Validate() error { return nil }`. This method is executed by ucfg to capture additional config failures and report these errors on Unpack. If error is reported, ucfg adds some additional context with path to hint which prospector config failed. In future ucfg might try to collect mutliple errors for reporting (think `-configtest`).

File: filebeat/crawler/prospector.go [hunk 4]
```diff
@@ -5,14 +5,17 @@ import (
...
 	cfg "github.com/elastic/beats/filebeat/config"
 	"github.com/elastic/beats/filebeat/harvester"
 	"github.com/elastic/beats/filebeat/input"
+	"github.com/elastic/beats/libbeat/common"
 	"github.com/elastic/beats/libbeat/logp"
 )
 
 type Prospector struct {
-	ProspectorConfig cfg.ProspectorConfig
+	ProspectorConfig prospectorConfig
```
Comment (urso) [inline filebeat/crawler/prospector.go]: maybe name field `config`. having `prospector.ProspectorConfig` is pretty stuttery.

File: filebeat/harvester/harvester.go [hunk 5]
```diff
@@ -37,8 +40,31 @@ type Harvester struct {
...
+	MaxBackoffDuration time.Duration
+	CloseOlder         string `config:"close_older"`
+	CloseOlderDuration time.Duration
+	ForceCloseFiles    bool                   `config:"force_close_files"`
+	ExcludeLines       []*regexp.Regexp       `config:"exclude_lines"`
+	IncludeLines       []*regexp.Regexp       `config:"include_lines"`
+	MaxBytes           int                    `config:"max_bytes"`
+	Multiline          *input.MultilineConfig `config:"multiline"`
+	JSON               *input.JSONConfig      `config:"json"`
+}
```
Comment (urso) [inline filebeat/harvester/harvester.go]: this struct needs some more cleanup?

File: filebeat/config/config.go (hunk 1 above)
Comment (ruflin) [inline filebeat/config/config.go]: Changed

File: filebeat/crawler/crawler.go (hunk 2 above)
Comment (ruflin) [inline filebeat/crawler/crawler.go]: I save that for a later PR ;-)

File: filebeat/crawler/prospector.go (hunk 3 above)
Comment (ruflin) [inline filebeat/crawler/prospector.go]: I added the method but currently left it empty. Config stuff was moved to config.go

File: filebeat/crawler/prospector.go (hunk 4 above)
Comment (ruflin) [inline filebeat/crawler/prospector.go]: Renamed to config and made private. Now that stuff is better separated it is also not that confusing anymore ;-)

File: filebeat/harvester/harvester.go (hunk 5 above)
Comment (ruflin) [inline filebeat/harvester/harvester.go]: Yes, and not only this one, but not in this PR (see note in PR).

**Comment by ruflin** [comment]
@urso Update pushed

**Pull request merged**